### PR TITLE
Add gob support

### PR DESCRIPTION
**Pull request by dascapytal1559**

### CODE REVIEW

**Comment by eapenkin** [comment]
Dear @dascapytal1559,

Please have a look at v0.1.25. It supports binary marshaling from/to packed BCD.

**Comment by dascapytal1559** [comment]
> Dear @dascapytal1559,
> 
> Please have a look at v0.1.25. It supports binary marshaling from/to packed BCD.

Yep nice - I've updated so gob uses binary marshall

**Comment by tafaquh-fiddin** [comment]
Dear @dascapytal1559 , I am able to do Marshaling without these changes after updating to version v0.1.33
```
var buf bytes.Buffer
	encoder := gob.NewEncoder(&buf)
	err := encoder.Encode(payload)
	if err != nil {
		return "", afferr.
			Code(core.ErrCodeInvalidSearchToken).
			StackTrace().
			Wrap(err, "failed marshal search token")
	}
```
I think this PR is no longer needed

**Comment by eapenkin** [comment]
Dear @dascapytal1559,

Here is an excerpt from `encoding/gob`:

> Gob can encode a value of any type implementing the [GobEncoder](https://pkg.go.dev/encoding/gob#GobEncoder) or [encoding.BinaryMarshaler](https://pkg.go.dev/encoding#BinaryMarshaler) interfaces by calling the corresponding method, in that order of preference.
> Gob can decode a value of any type implementing the [GobDecoder](https://pkg.go.dev/encoding/gob#GobDecoder) or [encoding.BinaryUnmarshaler](https://pkg.go.dev/encoding#BinaryUnmarshaler) interfaces by calling the corresponding method, again in that order of preference.

As suggested by @tafaquh-fiddin, there is no need for GOB dedicated methods.

**Comment by dascapytal1559** [comment]
Yep cool, I'm happy to accept this. Is the JsonMashalling worth adding, if so I'll make a separate PR, if not I'll just delete this. 

Thanks guys.

File: decimal.go [hunk 1]
```diff
@@ -709,6 +709,29 @@ func (d Decimal) MarshalText() ([]byte, error) {
...
+	*d, err = Parse(unquoted)
+	return err
+}
+
+// MarshalJSON implements the json.Marshaler interface.
+func (d Decimal) MarshalJSON() ([]byte, error) {
+	return []byte("\"" + d.String() + "\""), nil
+}
+
+// GobDecode implements the gob.GobDecoder interface for gob serialization.
```
Comment (ccoVeille) [inline decimal.go]:
```suggestion
// GobDecode implements the [gob.GobDecoder] interface for gob serialization.
```

File: decimal.go [hunk 2]
```diff
@@ -709,6 +709,29 @@ func (d Decimal) MarshalText() ([]byte, error) {
 	return []byte(d.String()), nil
 }
 
+// UnmarshalJSON implements the json.Unmarshaler interface.
```
Comment (ccoVeille) [inline decimal.go]:
```suggestion
// UnmarshalJSON implements the [json.Unmarshaler] interface.
```

File: decimal.go [hunk 3]
```diff
@@ -709,6 +709,29 @@ func (d Decimal) MarshalText() ([]byte, error) {
...
+func (d *Decimal) UnmarshalJSON(s []byte) error {
+	var err error
+	unquoted := string(s[1 : len(s)-1])
+	*d, err = Parse(unquoted)
+	return err
+}
+
+// MarshalJSON implements the json.Marshaler interface.
```
Comment (ccoVeille) [inline decimal.go]:
```suggestion
// MarshalJSON implements the [json.Marshaler] interface.
```

File: decimal.go [hunk 4]
```diff
@@ -709,6 +709,29 @@ func (d Decimal) MarshalText() ([]byte, error) {
...
+func (d Decimal) MarshalJSON() ([]byte, error) {
+	return []byte("\"" + d.String() + "\""), nil
```
Comment (ccoVeille) [inline decimal.go]:
```suggestion
	return []byte(`"` + d.String() + `"`), nil
```

File: decimal.go [hunk 5]
```diff
@@ -709,6 +709,29 @@ func (d Decimal) MarshalText() ([]byte, error) {
...
+}
+
+// GobDecode implements the gob.GobDecoder interface for gob serialization.
+func (d *Decimal) GobDecode(data []byte) error {
+	return d.UnmarshalBinary(data)
+}
+
+// GobEncode implements the gob.GobEncoder interface for gob serialization.
```
Comment (ccoVeille) [inline decimal.go]:
```suggestion
// GobEncode implements the [gob.GobEncoder] interface for gob serialization.
```

File: decimal_test.go [hunk 6]
```diff
@@ -70,6 +88,198 @@ func TestDecimal_Interfaces(t *testing.T) {
 	}
 }
 
+func TestDecimal_UnmarshallJson(t *testing.T) {
```
Comment (ccoVeille) [inline decimal_test.go]:
```suggestion
func TestDecimal_UnmarshalJSON(t *testing.T) {
```

File: decimal_test.go [hunk 7]
```diff
@@ -70,6 +88,198 @@ func TestDecimal_Interfaces(t *testing.T) {
...
+			}
+
+			if data.Value.Cmp(tt.expected) != 0 {
+				t.Errorf("json.Unmarshal() failed, got=%q, want=%q", data.Value, tt.expected)
+			}
+		})
+	}
+}
+
+func TestDecimal_MarshallJson(t *testing.T) {
```
Comment (ccoVeille) [inline decimal_test.go]:
```suggestion
func TestDecimal_MarshalJSON(t *testing.T) {
```

File: decimal_test.go [hunk 8]
```diff
@@ -70,6 +88,198 @@ func TestDecimal_Interfaces(t *testing.T) {
...
+			}
+
+			if string(result) != tt.expected {
+				t.Errorf("json.Marshal() failed, got=%q, want=%q", string(result), tt.expected)
+			}
+		})
+	}
+}
+
+func TestDecimal_JsonEnd2End(t *testing.T) {
```
Comment (ccoVeille) [inline decimal_test.go]:
```suggestion
func TestDecimal_JSONEnd2End(t *testing.T) {
```

**Pull request closed without merging**